### PR TITLE
Switch to TemporaryArrays while topologically sorting graphs

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Addressed 0.6% of allocations in a simple typing/lightbulb scenario:

image
image

This is a scenario where we need scratch space to store items for an algorithm, but we know the scratch space will never be more than 2 elements.  TemporaryArray excels here, reserving space on the stack for up to 4 elements, and spilling over to the heap for more than that.  Thsi means in all these cases, we produce no garbge and are effectively optimal in terms of space/cpu needed.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@333fred @RikkiGibson ptal

File: src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs:1581]: this guy was particularly egregious.  Constantly created arrays of just 1-2 elements long just to then walk and throw them away.  

These algorithms really benefit from just using a scratch buffer to write things into.  

If this wasn't a generic algorithm, and was known to be used *only* by the decisinon-dag stuff, we could optimize even further by just using a stack allocated span for this.  but i didn't want to go that far.

File: src/Compilers/CSharp/Portable/Binder/SwitchExpressionBinder.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/SwitchExpressionBinder.cs:122]: more array allocations just to pass back a singel successor item.

File: src/Compilers/Core/Portable/Collections/TopologicalSort.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/TopologicalSort.cs:20]:
```suggestion
```

File: src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs:1581]: Can we not use TemporaryArray<T> here? It has better cache performance than the object pools when the majority of cases do not exceed 4 elements.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs:1581]: these take in delegates that are called to add elements to the scratch space.  We'd need to pass temporary-array by-ref to make that work, and i was hesitant about defining a whole new delegate type for that.  i'm not opposed to it, but this was def path of least resistance.  And the builders themselves are only created once for the lifetime of the whole graph walk, so this seemed good enough (For now).

But i am open to changing this in this PR if we feel TempArray would be ncier.

File: src/Compilers/Core/Portable/Collections/TopologicalSort.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Collections/TopologicalSort.cs:50]: Not necessarily blocking the PR since I don't know how we're going to untangle the problem, but the disposal pattern for `TemporaryArray` was let in unintentionally into the compiler, so I hesitate to pile onto the problem before it can be resolved.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Collections/TopologicalSort.cs:50]: > but the disposal pattern for TemporaryArray was let in unintentionally into the compiler

I'm not sure what you mean here :)

This is a Disposable type, and there's nothign wrong with disposable types in the compiler layer (never has been).  We use them all over the place (for example, the parser) for very clean code that is safe to making mistakes (the sorts both you and i have found in recent PRs).

The Dispose concern existed for types that are also manually *Freeable*.  Because there was a concern of someone both disposing and freeing an item.  TempArray exposes no free mechanism *intentionally*.  So you can't accidentally fall into that path.  This is a core part of the design precisely so that it can fit in both IDE and compiler and be used properly with `using` without issue in either layer.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Collections/TopologicalSort.cs:50]: This is also why the PR that adds *Free* to TempArray is problematic.  It actively moves TempArray into having a duality that will actively make it easier to make mistakes with.  We do *not* want such a Duality, and instead want this to behave like normal Disposable types (including the other ones used in the compiler) that only have one clear form that is basically impossible to get wrong :)

Comment (RikkiGibson) [reply src/Compilers/Core/Portable/Collections/TopologicalSort.cs:50]: It feels unfortunate that in order to get "do something whenever we exit this block through a `return`" in C# we also have to opt-in to the overhead of "do something whenever we exit this block through a `throw`".

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Collections/TopologicalSort.cs:50]: Note: overhead appears to be... nothing/near-nothing? :)  As an example, tehse sorts of patterns are used extensively in the parser, especially in hotpaths around speculation.  I've tested with/without the using, and found the perf effectively identical.  Maybe this was a concern (or perhaps in hte netfx days?).  It doesn't seem to matter today.  Or if it does exist, it's so far dwarfed by the other work we do around this.

In cases like this, for example, we get such nice wins with data locality just using the stack and not producing garbage, that a simple try/finally seems irrelevant in comparison.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson @jcouv @jaredpar ptal :)

**Review by RikkiGibson — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@jcouv Ptal.  I switched to just explicit Dispose while the offlien convo about 'usings' happens.  We can keep this as is if we decide no-usings.  Or we can move to usings otherwise in a future PR.  But this is a nice win (And cleanup) that would be great to get in while that longer discussion happens.  Thanks!

**Comment by CyrusNajmabadi** [comment]
@jcouv reverted that as per offline discussion.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 7). From discussion with compiler team we want to unblock progress on this performance effort and will have separate discussion on usage of disposal pattern in the compiler. Thanks!

**Pull request merged**
